From oneukum@suse.de Tue Jun 12 06:36:31 2007
From: Oliver Neukum <oneukum@suse.de>
Date: Tue, 12 Jun 2007 15:36:07 +0200
Subject: USB: memory leak in iowarrior.c
To: Christian Lucht <lucht@codemercs.com>, Greg KH <gregkh@suse.de>, linux-usb-devel@lists.sourceforge.net
Message-ID: <200706121536.07551.oneukum@suse.de>
Content-Disposition: inline

this is a classical memory leak in the ioctl handler. The buffer is simply
never freed. This fixes it the obvious way.

Signed-off-by: Oliver Neukum <oneukum@suse.de>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/misc/iowarrior.c |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

--- a/drivers/usb/misc/iowarrior.c
+++ b/drivers/usb/misc/iowarrior.c
@@ -493,8 +493,8 @@ static int iowarrior_ioctl(struct inode 
 
 	/* verify that the device wasn't unplugged */
 	if (!dev->present) {
-		mutex_unlock(&dev->mutex);
-		return -ENODEV;
+		retval = -ENODEV;
+		goto error_out;
 	}
 
 	dbg("%s - minor %d, cmd 0x%.4x, arg %ld", __func__, dev->minor, cmd,
@@ -577,9 +577,10 @@ static int iowarrior_ioctl(struct inode 
 		retval = -ENOTTY;
 		break;
 	}
-
+error_out:
 	/* unlock the device */
 	mutex_unlock(&dev->mutex);
+	kfree(buffer);
 	return retval;
 }
 
